File: dsa/li04_twopointers/L283_MoveZeroes.java
package li04_twopointers;

import java.util.Arrays;

public class L283_MoveZeroes {

	public static void main(String[] args) {
		L283_MoveZeroes obj = new L283_MoveZeroes();
		int[] nums = new int[] { 0, 1, 0, 3, 12 };
		obj.moveZeroes(nums);
		System.out.println(Arrays.toString(nums));
	}

	// Problem: https://leetcode.com/problems/move-zeroes/
	// Idea: 1st pointer waits at non 0 place, 2nd pointer waits at 0 place, then swap.
	public void moveZeroes(int[] nums) {
		int r = 0, l = 0, n = nums.length;
		while (r < n && l < n) {
			while (r < n && nums[r] == 0)
				r++;
			while (l < n && nums[l] != 0 && l < r)
				l++;
			if (r < n && l < n) {
				int temp = nums[l];
				nums[l] = nums[r];
				nums[r] = temp;
				r++;
				l++;
			}
		}
	}

}

File: dsa/concepts/04_twopointers/python/l283_move_zeroes.py
"""
L283 - Move Zeroes
Problem: https://leetcode.com/problems/move-zeroes/
Idea: 1st pointer waits at non 0 place, 2nd pointer waits at 0 place, then swap.
Time: O(n)
Space: O(1)
"""

def move_zeroes(nums):
    """
    Move all zeroes to the end while maintaining relative order of non-zero elements.
    
    Args:
        nums: List[int] - Array of integers
    
    Returns:
        None - Modifies the array in-place
    """
    n = len(nums)
    r = l = 0
    
    while r < n and l < n:
        # Find next non-zero element
        while r < n and nums[r] == 0:
            r += 1
        # Find next zero element that comes before r
        while l < n and nums[l] != 0 and l < r:
            l += 1
        
        if r < n and l < n:
            nums[l], nums[r] = nums[r], nums[l]
            r += 1
            l += 1


if __name__ == "__main__":
    nums = [0, 1, 0, 3, 12]
    move_zeroes(nums)
    print(nums)  # Expected: [1, 3, 12, 0, 0]
